# COSADAMA Introduction to Data Science Study

- 작성자: 정 찬
- 참고자료: 2020년 코사다마 데이터 분석 커리큘럼(박하람), 2021년 코사다마 데이터 분석 커리큘럼(조용주), 파이썬 머신러닝 완벽 가이드(권철민), 비주얼 프로그래밍 수업(김승태), 파이썬으로 데이터 주무르기(민형기), 

Introduction to Data Science 커리큘럼의 목적은 데이터 '수집, '조작', '시각화'입니다. 1~2주에는 이미 만들어진 데이터를 조작하고, 효과적으로 시각화 하는 방법에 대해 배울 예정입니다.

이번 주에는 **데이터 분석 모듈 Pandas, 수치 데이터 처리 모듈 Numpy, 시각화 모듈 matplitlib**에 대해 배웁니다. 앞으로 가장 많이 사용하게 되실 모듈이니 꼼꼼히 공부하시고, 스스로 사용해보시며 친해지시길 바랍니다.

## 1. Pandas

모든 데이터, 정보가 줄글로 씌여 있다고 생각해 봅시다. 예컨대 
> 사과 100개, 귤 200개, 양초 50개가 있습니다.

이정도의 데이터는 한 눈으로 봐도 이해가 될 수 있습니다. 하지만 1960년대부터 2022년까지 매일, 모든 지역의, 모든 과일의 시세를 직관적으로 파악하기는 힘듭니다. 그래서 사람들은 대부분 **행row과 열column**으로 된 표table 형태의 2차원 데이터를 사용합니다. Excel이나 Numbers를 떠올리셨다면 정답입니다. **Pandas**는 이러한 2차원 데이터를 가공하고 처리하기 위해 만들어졌습니다. 

그렇다면 왜 Excel을 사용하지 않냐는 물음이 생길 수 있습니다. 우리가 다루는 데이터들은 하나하나 확인하기 어려울만큼 거대합니다. 컬럼이 한 화면에 들어오기 어려운 경우도 많고, 애초에 데이터가 많아서 Excel이 실행되지 않기도 합니다. 그래서 판다스는 보다 개념적인 명령을 통해 큰 데이터의 추출, 조작을 도와줍니다.

Excel을 쓸 줄 아는 사람들은 많지만, Excel의 모든 기능을 사용할 줄 아는 전문가는 많지 않습니다. Pansdas도 마찬가지 입니다. 많이 사용되는 기능 위주로 작성했으니 더 필요한 부분은 구글에게 물어가며 이후 프로젝트를 진행해 봅시다.

### 1.1. Pandas의 자료 구조

- **Series**: 1차원 배열의 자료구조. 하나의 열(column컬럼)으로 이루어짐. 인덱스와 함께 구성됨. 
- **DataFrame**: 2차원 배열의 자료구조. 여러 열로 이루어짐.

Series와 DataFrame의 가장 큰 차이는 Series는 칼럼이 하나, DataFrame은 칼럼이 여러 개!

> 행.. 열..?

> 보통 가로, 세로 순서로 말하죠? 행과 열도 가로(행row), 세로(열column) 순서입니다.

> 하지만 엄밀하게 말해서 "행렬"Matrix과 같은 의미는 아닙니다. 연산 규칙이 달라지거든요. 일단 DataFrame이 행렬Matrix보다 더 직관적이라고만 알고 계셔도 충분합니다.

> 배열(array): 같은 타입의 변수들로 이루어진 유한 집합. 


In [ ]:
# 파이썬 3.8버전 추가하기(숙지할 필요 없음)

import sys 
sys.path.append("/usr/local/lib/python3.8/site-packages")

In [2]:
# pandas 모듈 import
# pandas 모듈을 가져와! 
# 코드에서는 pd라고 축약해서 쓸거야!
import pandas as pd

In [3]:
# Series
obj = pd.Series([4,7,-5,3])   # 왼쪽의 인덱스와 결합됨. 오른쪽이 시리즈의 값.
obj

0    4
1    7
2   -5
3    3
dtype: int64

In [4]:
obj.index

# 숫자 범위로 된 인덱스. 0부터 4까지 1씩 증가.

RangeIndex(start=0, stop=4, step=1)

In [5]:
# 인덱스 지정하기
obj2 = pd.Series([4,7,-5,3], index=['d','b','a','c'])
obj2   # 항상 어떤 데이터가 들어있는지 확인해보는 것 잊지말기!

d    4
b    7
a   -5
c    3
dtype: int64

In [6]:
# 인덱스가 'd'인 값 정수 6으로 바꾸기
obj2['d'] = 6
obj2

d    6
b    7
a   -5
c    3
dtype: int64

In [7]:
# 딕셔너리에서 Series 만들기 
sdata = {'Ohio': 35000, 'Texas':71000,
        'Oregon':16000, 'Utah':5000}
obj3 = pd.Series(sdata)
obj3

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

In [8]:
# 리스트로 인덱스 설정하기
state = ['California','Ohio','Oregon','Texas']
obj4 = pd.Series(sdata, index=state)
obj4

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64

- NaN: Not a Number -> 값이 없다!

     -  문제: Series나 DataFrame을 열었는데 공백으로 표시된다면 그 데이터는 NaN일까?

    - 정답: NO! NaN은 NaN이라고 분명하게 표시됩니다. 확인해 봐야 알겠지만, ''이나 ' ' 같은 모양으로 된 "데이터가 들어가 있"을 가능성이 높아요. 물론 0도 0이라고 분명하게 표시해 줍니다.

### DataFrame
- 구성: 인덱스index, 열column, 데이터
- 여러 행, 여러 열로 이루어진 2차원 데이터 구조체
- DataFrame == 여러 개의 Series
- Series는 DataFrame의 열column이 될 수 있음

In [9]:
# 딕셔너리를 이용해 DataFrame 생성하기 
data = {'state':['Ohio','Ohio','Ohio','Nevada','Nevada'],
       'year':[2000,2001,2002,2001,2002],
       'pop':[1.5,1.7,3.6,2.4,2.9]}
df = pd.DataFrame(data)
df

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9

In [10]:
# 칼럼을 원하는 순서대로 보기 
df2 = pd.DataFrame(data, columns=['year','state','pop'])
df2

year   state  pop
0  2000    Ohio  1.5
1  2001    Ohio  1.7
2  2002    Ohio  3.6
3  2001  Nevada  2.4
4  2002  Nevada  2.9

In [11]:
# 인덱스를 지정하고 싶다면
df3 = pd.DataFrame(data, columns=['year','state','pop','debt'],
                  index=['one','two','three','four','five'])
df3

year   state  pop debt
one    2000    Ohio  1.5  NaN
two    2001    Ohio  1.7  NaN
three  2002    Ohio  3.6  NaN
four   2001  Nevada  2.4  NaN
five   2002  Nevada  2.9  NaN

### 1.2. 데이터 준비

1. [공공데이터포털]('https://www.data.go.kr/data/15044251/fileData.do)에서 서울교통공사_월별 승차인원 csv 파일 다운로드
2. 특정 저장위치(예: "D:/COSADAMA/Week1/train.csv")에 저장하기
3. 2번 예시처럼 저장 위치 복사해 두기
5. pd.read_csv('저장위치/파일이름.csv', encoding = 'utf-8', sep = ',', names = ['원하는','컬럼 이름으로', '변경하기'], usecols = ['원본 데이터의', '원하는', '컬럼이름들'], index_col = '인덱스로 설정하고 싶은 컬럼명')

- encoding: 한글, 일본어, 중국어 등 영어 외의 언어를 읽어올 때 필요한 인코딩 설정
- sep: 데이터가 어떤 방식으로 나뉘는지.
    - csv: comma-separated values -> 컴마로 구분된 데이터 -> sep = ','
    - tsv: tab-separated values -> tab으로 구분된 데이터 -> sep = '\t'
    - txt: ','일 수도, '\t'일 수도, '|'일 수도, 그 외의 데이터일 수도 있음 -> 직접 확인
- names: 전체 컬럼의 이름을 원하는 대로 다시 입력하기
- usecols: 필요한 컬럼만 가져오기
- index_col: 특정 컬럼을 인덱스로 설정하기

In [14]:
############ 나중에 본인의 경로/train.csv로 바꾸기
subway_df = pd.read_csv('C:/Users/jung6/Downloads/서울교통공사_월별 승차인원_20211130.csv')
subway_df.head()  # 데이터가 너무 많을 때 앞의 5개만 불러오기

����  ȣ��  ����ȣ      ����.1  2021��1��  2021��2��  2021��3��  2021��4��  \
0     1    1    150     ���￪(1)   863461.0   890851.0    1827089    1841712   
1     2    1    151      ��û(1)   420304.0   420686.0     954481     955202   
2     3    1    152        ����   685203.0   706115.0    1598997    1572365   
3     4    1    153  ����3��(1)   507414.0   549933.0    1084476    1059282   
4     5    1    154     ����5��   469264.0   475339.0    1100841    1071587   

   2021��5��  2021��6��  2021��7��  2021��8��  2021��9��  2021��10��  \
0    1841724    1152743     973900     942249     957461     1125594   
1     902354     536217     477840     451248     451677      507332   
2    1522428     885428     767958     720164     702153      799319   
3    1053215     691110     612154     601553     608293      696209   
4    1057283     600243     536493     509050     519447      583177   

   2021��11��  
0     1270865  
1      590817  
2      919350  
3      712182  
4      595547

앗! 한글이 모두 깨져서 나오네요! 이럴 때는 **인코딩encoding**을 설정해 주시면 됩니다. 컴퓨터는 0과 1만 인식하기 때문에 우리가 사용하는 한글, 영어를 인식할 때 특정 문자 집합과 숫자를 매핑한 문자 집합Character Set을 사용합니다. 이때 **문자를 byte 코드로 표현하는 작업을 인코딩encoding**이라고 합니다.

대표적인 한글 인코딩에는 유니코드(**utf-8**, utf-16, utf-32), **euc-kr**, **cp949** 등이 있습니다.


In [16]:
# 인코딩 지정해서 읽어오기
subway_df = pd.read_csv('C:/Users/jung6/Downloads/서울교통공사_월별 승차인원_20211130.csv',
                        encoding = 'euc-kr')
subway_df.head()  # 데이터가 너무 많을 때 앞의 5개만 불러오기

연번  호선  역번호       역명   2021년1월   2021년2월  2021년3월  2021년4월  2021년5월  \
0   1   1  150   서울역(1)  863461.0  890851.0  1827089  1841712  1841724   
1   2   1  151    시청(1)  420304.0  420686.0   954481   955202   902354   
2   3   1  152       종각  685203.0  706115.0  1598997  1572365  1522428   
3   4   1  153  종로3가(1)  507414.0  549933.0  1084476  1059282  1053215   
4   5   1  154     종로5가  469264.0  475339.0  1100841  1071587  1057283   

   2021년6월  2021년7월  2021년8월  2021년9월  2021년10월  2021년11월  
0  1152743   973900   942249   957461   1125594   1270865  
1   536217   477840   451248   451677    507332    590817  
2   885428   767958   720164   702153    799319    919350  
3   691110   612154   601553   608293    696209    712182  
4   600243   536493   509050   519447    583177    595547

서울교통공사는 월별 승차인원 데이터를 euc-kr 형식으로 저장했군요! 저장할 때 어떤 인코딩encoding 형식을 사용했는지에 따라 데이터를 불러올 때 확인하는 작업이 필요하답니다.

역번호는 필요 없는 상황일 때 연번, 호선, 역명, 1 ~ 3월의 데이터를 가져오고, 연번이 중복되지 않는 값이니 index로 만들어 봅시다.

In [25]:

subway_df = pd.read_csv('C:/Users/jung6/Downloads/서울교통공사_월별 승차인원_20211130.csv',
                        encoding = 'euc-kr', 
                        usecols = ['연번', '호선', '역명', 
                                   '2021년1월',	'2021년2월', '2021년3월'],
                        index_col = '연번')

subway_df.head()  # 데이터가 너무 많을 때 앞의 5개만 불러오기

호선       역명   2021년1월   2021년2월  2021년3월
연번                                          
1    1   서울역(1)  863461.0  890851.0  1827089
2    1    시청(1)  420304.0  420686.0   954481
3    1       종각  685203.0  706115.0  1598997
4    1  종로3가(1)  507414.0  549933.0  1084476
5    1     종로5가  469264.0  475339.0  1100841

### 1.3. 데이터 확인

|함수|활용방법|
|----|--------|
|df.type()|데이터 형식 파악하기|
|df.head()|처음 n개의 데이터 보이기. 기본 설정은 5|
|df.tail()|마지막 n개의 데이터 보이기. 기본 설정은 5|
|df.shape|몇 행, 몇 열로 구성된 데이터인지 확인|
|df.info()|DataFrame의 총 데이터 수, 컬럼별 데이터 타입, Null 개수|
|df.describe()|DataFrame의 통계 요약 정보. 숫자형 컬럼의 count, mean, std, min, 25%, 50%, 75%, max 값 표시|
|df[].value_counts()|특정 컬럼에 속한 데이터가 각각 몇개 속해 있는지 확인. 예를 들어 표의 등급을 의미하는 Pclass는 1, 2, 3으로 이루어져 있음|
|df.sort_values(by='')|특정 컬럼 값을 기준으로 DataFrame 정렬하기|

1. [Kaggle](https://www.kaggle.com/c/titanic/data)에 가입하기
2. Data Explorer의 train.csv 다운로드 하기
3. 특정 저장 위치(예: "D:/COSADAMA/Week1/train.csv")에 저장하기
4. 2번 예시처럼 저장 위치 복사해 두기
5. pd.read_csv('저장위치/파일이름.csv', encoding = 'utf-8', sep = ',', names = ['원하는','컬럼 이름으로', '변경하기'], usecols = ['원본 데이터의', '원하는', '컬럼이름들'])

In [27]:
############ 나중에 본인의 경로/train.csv로 바꾸기
titanic_df = pd.read_csv('C:/Users/jung6/Downloads/train.csv')
titanic_df.head()  # 데이터가 너무 많을 때 앞의 5개만 불러오기

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [30]:
type(titanic_df)

pandas.core.frame.DataFrame

In [28]:
# 뒤의 5개만 가져오기
titanic_df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [29]:
# 앞의 10개만 
titanic_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [33]:
titanic_df.shape

(891, 12)

In [34]:
# DataFrame의 총 데이터 건수, 데이터 타입, Null 건수
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


- 0 to 890: 전체 row 수
- Non-Null Count: 각 컬럼에서 몇개의 데이터가 non-null인지. 전체 891개 중 714개만 존재한다는 건 분명 뭔가 missing data가 있다 의미!
- non-null: null값이 아닌 것.(값이 있다)
- object는 편하게 파이썬에서 배운 문자열이라고 생각하면 됩니다.

In [35]:
# DataFrame의 통계요약 정보 
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [44]:
# DataFrame 중 'Pclass' 칼럼만 떼어보기 
titanic_Pclass = titanic_df['Pclass']
titanic_Pclass

0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64

In [45]:
# 여러 칼럼을 떼어보기 
titanic_df[['Pclass','Age']]   # 여러 개 칼럼은 [ ]로 한번 더 감싸줍니다.

Pclass   Age
0         3  22.0
1         1  38.0
2         3  26.0
3         1  35.0
4         3  35.0
..      ...   ...
886       2  27.0
887       1  19.0
888       3   NaN
889       1  26.0
890       3  32.0

[891 rows x 2 columns]

In [37]:
# 특정 컬럼의 값이 얼마나 포함되어 있는지 확인
# 컬럼 값의 유형과 건수 확인
# 1등급 216, 2등급 184, 3등급 491석

titanic_df['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [43]:
titanic_df.sort_values(by='Pclass')

PassengerId  Survived  Pclass  \
445          446         1       1   
310          311         1       1   
309          310         1       1   
307          308         1       1   
306          307         1       1   
..           ...       ...     ...   
379          380         0       3   
381          382         1       3   
382          383         0       3   
371          372         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
445                          Dodge, Master. Washington    male   4.0      0   
310                     Hays, Miss. Margaret Bechstein  female  24.0      0   
309                     Francatelli, Miss. Laura Mabel  female  30.0      0   
307  Penasco y Castellana, Mrs. Victor de Satode (M...  female  17.0      1   
306                            Fleming, Miss. Margaret  female   NaN      0   
..                                                 ...     ...   ...    ...   
379                        Gustafsson, Mr. Karl Gideon    male  19.0      0   
381                        Nakid, Miss. Maria ("Mary")  female   1.0      0   
382                                 Tikkanen, Mr. Juho    male  32.0      0   
371                          Wiklund, Mr. Jakob Alfred    male  18.0      1   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch             Ticket      Fare Cabin Embarked  
445      2              33638   81.8583   A34        S  
310      0              11767   83.1583   C54        C  
309      0           PC 17485   56.9292   E36        C  
307      0           PC 17758  108.9000   C65        C  
306      0              17421  110.8833   NaN        C  
..     ...                ...       ...   ...      ...  
379      0             347069    7.7750   NaN        S  
381      2               2653   15.7417   NaN        C  
382      0  STON/O 2. 3101293    7.9250   NaN        S  
371      0            3101267    6.4958   NaN        S  
890      0             370376    7.7500   NaN        Q  

[891 rows x 12 columns]

### 1.4. 원하는 데이터 선택

|데이터 선택 방법|형태|활용방법|해석|
|----------------|----|--------|----|
|슬라이싱slicing|df[['원하는', '컬럼명']]|titanic_df[['Pclass','Age']]|Pclass, Age 컬럼 가져와|
|boolean indexing|df[] == ?|titanic_df[(titanic_df['Pclass']==1)\|(titanic_df['Age']<20)]|Pclass==1이고, Age<20인 데이터 가져와|
|loc[] | df.loc[['인덱스'], ['열']]|titanic_df.loc[:10,'PassengerId':'Age']|처음 10개행의 Passengerld부터 Age 까지 값 가져와|
|iloc[]| df.iloc[인덱스 값, 열 값]|titanic_df.iloc[:,2:5]|전체 인덱스의 2부터 4행 데이터 가져와|

In [46]:
# 슬라이싱
titanic_df[['Pclass','Age']]

Pclass   Age
0         3  22.0
1         1  38.0
2         3  26.0
3         1  35.0
4         3  35.0
..      ...   ...
886       2  27.0
887       1  19.0
888       3   NaN
889       1  26.0
890       3  32.0

[891 rows x 2 columns]

In [50]:
# 불린 인덱싱
titanic_df[(titanic_df['Pclass']==1)|(titanic_df['Age']<20)]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
7              8         0       3   
9             10         1       2   
..           ...       ...     ...   
875          876         1       3   
877          878         0       3   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
877                               Petroff, Mr. Nedelio    male  19.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare Cabin Embarked  
1        0  PC 17599  71.2833   C85        C  
3        0    113803  53.1000  C123        S  
6        0     17463  51.8625   E46        S  
7        1    349909  21.0750   NaN        S  
9        0    237736  30.0708   NaN        C  
..     ...       ...      ...   ...      ...  
875      0      2667   7.2250   NaN        C  
877      0    349212   7.8958   NaN        S  
879      1     11767  83.1583   C50        C  
887      0    112053  30.0000   B42        S  
889      0    111369  30.0000  C148        C  

[359 rows x 12 columns]

In [72]:
# df.loc[]
titanic_df.loc[:,'PassengerId':'Age']

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  
0                              Braund, Mr. Owen Harris    male  22.0  
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0  
2                               Heikkinen, Miss. Laina  female  26.0  
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  
4                             Allen, Mr. William Henry    male  35.0  
..                                                 ...     ...   ...  
886                              Montvila, Rev. Juozas    male  27.0  
887                       Graham, Miss. Margaret Edith  female  19.0  
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN  
889                              Behr, Mr. Karl Howell    male  26.0  
890                                Dooley, Mr. Patrick    male  32.0  

[891 rows x 6 columns]

In [73]:
# df.iloc[]
titanic_df.iloc[:,2:5]

Pclass                                               Name     Sex
0         3                            Braund, Mr. Owen Harris    male
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female
2         3                             Heikkinen, Miss. Laina  female
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female
4         3                           Allen, Mr. William Henry    male
..      ...                                                ...     ...
886       2                              Montvila, Rev. Juozas    male
887       1                       Graham, Miss. Margaret Edith  female
888       3           Johnston, Miss. Catherine Helen "Carrie"  female
889       1                              Behr, Mr. Karl Howell    male
890       3                                Dooley, Mr. Patrick    male

[891 rows x 3 columns]

### 1.5. 연산Aggregation 함수

- min(), max(), sum(), count()
- groupby()

### 1.6. 데이터 보간

- 결측치 확인: isna()
- 결측치 대체:
    - fillna()
    - replace()
- 인덱스 초기화: reset_index()

### 1.7. 데이터 삭제

- drop()
- del()

### 1.8. 사용자 지정함수 apply lambda()

### 1.9. DataFrame 병합

- pd.concat()
- pd.merge()


### 1.10. 기타

- 컬럼 이름 바꾸기: df.column = []
- 자료형 바꾸기: 
    - df.astype()
    - apply(pd.to_numeric)
- 로컬 파일의 특정 형식(csv) 파일 모두 읽어오기:
    - from os import listdir
    - from glob import iglob
- DataFrame을 csv, excel 형태로 로컬에 저장하기
    

## 2. NumPy

**NumPy**는 선형대수 기반 데이터/수치 분석 모듈입니다. 파이썬에서 여러 데이터를 연산할 때 for이나 while을 사용했습니다. 이런 형태를 loop라고 하고 계속해서 순차적으로 데이터를 확인하고 연산합니다. CPU를 사용해 하나하나 처리하는 형식입니다. 

하지만 numpy는 수치(숫자)로 된 데이터를 배열Array형식으로 인식하여 계산하기 때문에 기존보다 더 빠르게 만들어줍니다. 배열은 쉽게 말하면 n차원의 행렬matrix로 된 데이터입니다. NumPy가 빠른 이유는 GPU를 사용한 배열의 병렬 계산이 가능하기 때문입니다. (물론 cpu로도 가능합니다)

이번 커리큘럼의 목적은 "Python을 통한 데이터 수집/처리/시각화를 할 수 있다"이기 때문에 더 효율적인 연산에 대한 설명은 각설합니다.

앞선 설명과 같이 NumPy는 Pandas의 처리를 돕는 도구이지, 분석도구는 아닙니다. NumPy 또한 내용이 방대합니다. 기초적인 데이터 분석에 필요한 내용만 추렸으니 더 필요하다면 구글, 책을 통한 스터디를 추천드립니다.

### 2.1. 배열 생성 

### 2.2. ndarray 속성

### 2.3. 수학/ 통계 매서드

- np.sum(): 전체 또는 특정 축의 원소의 합계
- np.mean(): 산술평균
- np.std(): 표준편차
- np.var(): 분산
- np.min(), max(): 최소값, 최대값
- np.cumsum(): 누적합
- np.comprod(): 누적곱

### 2.4. array 선택: Slicing

- []
- 불린 인덱싱

### 2.5. ndarray, 난수 생성

- np.arange(): range의 numpy 버전. 0부터 함수 인자 값 -1까지의 값을 순차적으로 ndarray의 데이터값으로 변환해줌.
- np.zeros(): 함수 인자로 튜플 형태의 shape 값을 입력하면 모든 값을 0으로 채운 해당 shape를 가진 ndarray를 반환함.
- np.ones(): 함수 인자로 튜플 형태의 shape 값을 입력하면 모든 값을 1로 채운 해당 shape을 가진 ndarray를 반환함.

- np.random.rand(): 균등 분포에서 추철
- np.random.randint(): 주어진 범위 내의 임의의 난수
- np.random.randn(): 표준편차가 1이고 평균 0인 정규 분포에서 추출
- np.random.normal(): 정규 분포에서 추출

## 3. Matplotlib

우리는 이제 **시각화**를 통해 데이터를 직관적으로 표현할 겁니다. 판다스를 통해 데이터를 예쁘게 만들었어도, 처음 보는 사람이 코드부터 데이터프레임까지 확인하기는 어렵습니다. 그래서 우리는 시각화를 통해 점진적 증가나 하락, A와 B의 차이, 데이터의 분포 등을 고안해 냈습니다.

**Matplotlib**은 가장 간단한 시각화 도구 중 하나입니다. Seaborn, plotly도 있지만 가장 기본적이고 모든 내용을 포괄하기 때문에 Matplotlib을 선정했습니다.

Matplotlib 또한 모든 내용을 외운다고 생각하지 마세요. 어떤 작업을 할 수 있고, 어떤 매커니즘으로 코드를 작성하는지 흐름을 파악하시면 충분합니다. 어차피 막상 프로젝트 들어가면 다시 찾아보게 되니까요:)

### 3.0. 기본 설정

- plt.subplot(): 여러 개의 그래프 한 화면에 그리기
- plt.legend(): 범례 그리기
- plt.title(): 제목
- plt.xlabel(): x축 제목
- plt.ylabel(): y축 제목
- plt.grid(): 그리드 설정
- plt.xlim(), plt.ylim(): x, y축 범위 설정
- 스타일 지정: plt.plot() 괄호 내부에 color='', markers='', linestyle=''을 통해 색, 점 모양, 선 스타일 지정


### 3.1. 1차원

- 히스토그램
- pie chart
- bar chart

### 3.2. 2차원

- line plot
- scatter chart

### 3.3. 3차원(참고)

- bubble chart
- color map
- surface plot